File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <deque>
#include <stack>
#include <iostream>

using namespace std;

int solution(string s) {
    int answer = 0;
    deque<char> deq;
    string change = s;
    
    if(s.size() % 2 == 1) return 0;
    
    for(int i=0; i<s.size(); i++){
        
        string ch = "";
        
        for(int j=i; j<s.size(); j++){
            ch += change[j];
        }
        
        for(int j=0; j<i; j++){
            ch += change[j];
        }
        
        stack<char> st;
        stack<char> en;
        
        for(int j=0; j<s.size(); j++){
            st.push(ch[j]);
        }
        
        for(int j=0; j<s.size()/2; j++){
            char c = st.top();
            
            if(st.top() == '}' || st.top() == ')' || st.top() == ']'){
                en.push(c);
                st.pop();
            } else {
                break;
            }
            
            while(!en.empty()){
                if(en.top() == '}' && st.top() == '{'){
                    en.pop();
                    st.pop();
                } else if(en.top() == ')' && st.top() == '('){
                    en.pop();
                    st.pop();
                } else if(en.top() == ']' && st.top() == '['){
                    en.pop();
                    st.pop();
                } else break;
            }
        }
        
        if(st.empty() && en.empty()) answer++;
    }
    
    return answer;
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    public int solution(String s) {
        int answer = 0;
        
        int n = s.length();
        s += s;
        
        for(int i=0; i<n; i++){
            ArrayDeque<Character> st = new ArrayDeque<>();
            boolean isValid = true;
            
            for(int j=i; j<i+n; j++){
                
                if(s.charAt(j) == '{' || s.charAt(j) == '[' || s.charAt(j) == '('){
                    st.push(s.charAt(j));
                } else{
                    if(!st.isEmpty()){
                        if(st.peek() == '{' && s.charAt(j) == '}'){
                            st.pop();
                        } else if(st.peek() == '[' && s.charAt(j) == ']') {
                            st.pop();
                        } else if(st.peek() == '(' && s.charAt(j) == ')') {
                            st.pop();
                        } else {
                            isValid = false;
                            break;
                        }
                    } else {
                        isValid = false;
                        break;
                    }
                }
            }
            if (isValid && st.isEmpty()) {
                    answer++;
                }
            
        }
        
        return answer;
    }
}